File: tests/node/test_vc_smoke.mjs
// Voice-conversion bindings — hermetic smoke tests.
//
// Always-on (no env flag, no model downloads): every test in this file
// exercises only the typed surface of the RVC backend — option shapes,
// factory wiring, modelId round-trips. Tests that actually pull RVC
// weights or hit `$BLAZEN_RVC_VOICE_DIR/<voice>/` are gated behind:
//
//     BLAZEN_TEST_VC=1
//
// Build with the audio-vc-rvc feature enabled first:
//     pnpm --filter blazen run build
// (the canonical build uses `local-all` which includes `audio-vc-rvc`).
//
// Run:
//     pnpm exec ava tests/node/test_vc_smoke.mjs

import test from "ava";

import {
  RvcBackend,
  VcModel,
  VcError,
  VcEngineNotAvailableError,
  VcModelLoadError,
  VcConversionError,
  VcVoiceNotFoundError,
  VcUnsupportedError,
  VcIoError,
} from "../../crates/blazen-node/index.js";

const BLAZEN_TEST_VC = process.env.BLAZEN_TEST_VC;

const hasRvcExport = typeof RvcBackend === "function";
const hasVcModelExport = typeof VcModel === "function";

// Hermetic surface tests (run unconditionally — no weights, no network).
const H = (hasRvcExport && hasVcModelExport) ? test : test.skip;

// Live tests that need weights — gated on BLAZEN_TEST_VC.
const L = BLAZEN_TEST_VC ? test : test.skip;

// ---------------------------------------------------------------------------
// RvcBackend
// ---------------------------------------------------------------------------

H("vc · RvcBackend.create({}) builds a handle with modelId 'rvc'", (t) => {
  const backend = RvcBackend.create({});
  t.truthy(backend, "expected an RvcBackend instance");
  t.is(backend.modelId, "rvc");
});

H("vc · RvcBackend.create() with no options defaults to modelId 'rvc'", (t) => {
  const backend = RvcBackend.create();
  t.is(backend.modelId, "rvc");
});

H("vc · RvcBackend.create accepts topK / retrievalBlend / rvcVersion", (t) => {
  const backend = RvcBackend.create({
    topK: 16,
    retrievalBlend: 0.5,
    rvcVersion: "v1",
  });
  t.is(backend.modelId, "rvc");
});

H("vc · RvcBackend exposes convertVoice + streamConvertPcm as functions", (t) => {
  const backend = RvcBackend.create({});
  t.is(typeof backend.convertVoice, "function");
  t.is(typeof backend.streamConvertPcm, "function");
  t.is(typeof backend.listTargetVoices, "function");
  t.is(typeof backend.registerTargetVoice, "function");
});

H("vc · RvcBackend.convertVoice with a missing path rejects", async (t) => {
  const backend = RvcBackend.create({});
  const err = await t.throwsAsync(() =>
    backend.convertVoice("/nonexistent/path/that-does-not-exist.wav", "speaker-01"),
  );
  // Real-mode: voice lookup fires first and surfaces VcVoiceNotFoundError
  // (no voice profile registered for "speaker-01"). I/O happens after.
  // Stub-mode: VcEngineNotAvailableError.
  t.true(
    err instanceof VcError,
    `expected a VcError subclass, got ${err?.constructor?.name}: ${err?.message}`,
  );
});

H("vc · RvcBackend.registerTargetVoice surfaces VcUnsupportedError in real mode (or VcEngineNotAvailableError in stub mode)", async (t) => {
  const backend = RvcBackend.create({});
  const err = await t.throwsAsync(() =>
    backend.registerTargetVoice("new-speaker", "/nonexistent/ref.wav"),
  );
  t.true(
    err instanceof VcUnsupportedError || err instanceof VcEngineNotAvailableError,
    `expected VcUnsupportedError or VcEngineNotAvailableError, got ${err?.constructor?.name}: ${err?.message}`,
  );
});

H("vc · RvcBackend.listTargetVoices returns an array (or VcEngineNotAvailableError in stub mode)", async (t) => {
  const backend = RvcBackend.create({});
  try {
    const voices = await backend.listTargetVoices();
    // Real mode: returns either an empty array (no voice dir) or whatever
    // the user has registered under $BLAZEN_RVC_VOICE_DIR.
    t.true(Array.isArray(voices));
  } catch (err) {
    // Stub mode: VcEngineNotAvailableError.
    t.true(
      err instanceof VcEngineNotAvailableError,
      `expected VcEngineNotAvailableError, got ${err?.constructor?.name}: ${err?.message}`,
    );
  }
});

// ---------------------------------------------------------------------------
// VcModel aggregator
// ---------------------------------------------------------------------------

H("vc · VcModel.rvc({}) reports the same modelId as RvcBackend.create({})", (t) => {
  const direct = RvcBackend.create({});
  const unified = VcModel.rvc({});
  t.is(unified.modelId, direct.modelId);
});

H("vc · VcModel exposes the full convert + stream + voice-management surface", (t) => {
  const m = VcModel.rvc({});
  t.is(typeof m.convertVoice, "function");
  t.is(typeof m.streamConvertPcm, "function");
  t.is(typeof m.listTargetVoices, "function");
  t.is(typeof m.registerTargetVoice, "function");
});

H("vc · VcModel.rvc propagates the topK / retrievalBlend / rvcVersion options", (t) => {
  // Construction must succeed and report the same modelId regardless of opts.
  t.is(VcModel.rvc({ topK: 4 }).modelId, "rvc");
  t.is(VcModel.rvc({ retrievalBlend: 1.0 }).modelId, "rvc");
  t.is(VcModel.rvc({ rvcVersion: "v2" }).modelId, "rvc");
});

// ---------------------------------------------------------------------------
// Error classes are exported and form the expected hierarchy
// ---------------------------------------------------------------------------

H("vc · all Vc* error classes are exported", (t) => {
  t.is(typeof VcError, "function");
  t.is(typeof VcEngineNotAvailableError, "function");
  t.is(typeof VcModelLoadError, "function");
  t.is(typeof VcConversionError, "function");
  t.is(typeof VcVoiceNotFoundError, "function");
  t.is(typeof VcUnsupportedError, "function");
  t.is(typeof VcIoError, "function");
});

H("vc · Vc* subclasses all extend VcError extends Error", (t) => {
  t.true(VcEngineNotAvailableError.prototype instanceof VcError);
  t.true(VcModelLoadError.prototype instanceof VcError);
  t.true(VcConversionError.prototype instanceof VcError);
  t.true(VcVoiceNotFoundError.prototype instanceof VcError);
  t.true(VcUnsupportedError.prototype instanceof VcError);
  t.true(VcIoError.prototype instanceof VcError);
  t.true(VcError.prototype instanceof Error);
});

// ---------------------------------------------------------------------------
// Live tests (gated on BLAZEN_TEST_VC=1) — these need a registered voice
// under $BLAZEN_RVC_VOICE_DIR/<voice>/ to actually do anything useful.
// ---------------------------------------------------------------------------

L("vc [live] · RvcBackend.listTargetVoices returns whatever is registered", async (t) => {
  const backend = RvcBackend.create({});
  const voices = await backend.listTargetVoices();
  t.true(Array.isArray(voices));
  for (const v of voices) {
    t.is(typeof v.id, "string");
    t.is(typeof v.sampleRateHz, "number");
  }
});

File: tests/python/test_vc_smoke.py
"""Smoke tests for the `VcModel` / `VcChunk` / `VcStream` / `TargetVoice` surface.

The construction-free smoke (`test_vc_module_smoke`) always runs and
verifies the typed surface is reachable from Python — it requires only
that blazen-py was built with the `audio-vc-rvc` feature.

The live `RVC convert_voice` test is opt-in via `BLAZEN_RUN_RVC_TESTS=1`
plus `BLAZEN_RVC_VOICE_ID` and `BLAZEN_RVC_INPUT_WAV` because it depends
on a pre-staged voice profile under `$BLAZEN_RVC_VOICE_DIR` and a source
utterance on disk.
"""

import os

import pytest
import blazen as bz

pytestmark = pytest.mark.asyncio


def test_vc_module_smoke():
    """Construction-free smoke: types are importable when the
    `audio-vc-rvc` feature is compiled in."""
    if not hasattr(bz, "VcModel"):
        pytest.skip("blazen built without audio-vc-rvc feature")
    assert hasattr(bz, "VcChunk")
    assert hasattr(bz, "VcStream")
    assert hasattr(bz, "VcModel")
    assert hasattr(bz, "TargetVoice")
    assert hasattr(bz, "RvcError")
    # RvcError is a subclass of ProviderError.
    assert issubclass(bz.RvcError, bz.ProviderError)


async def test_vc_list_empty_dir(tmp_path):
    """`list_target_voices` returns an empty list when `voice_dir`
    contains no profiles."""
    if not hasattr(bz, "VcModel"):
        pytest.skip("blazen built without audio-vc-rvc feature")
    model = bz.VcModel.rvc(voice_dir=tmp_path)
    voices = await model.list_target_voices()
    assert isinstance(voices, list)
    assert voices == []


async def test_vc_register_voice_unsupported(tmp_path):
    """`register_target_voice` raises UnsupportedError on the current
    RVC backend (voice profiles must be pre-staged on disk)."""
    if not hasattr(bz, "VcModel"):
        pytest.skip("blazen built without audio-vc-rvc feature")
    model = bz.VcModel.rvc(voice_dir=tmp_path)
    ref_path = tmp_path / "reference.wav"
    ref_path.write_bytes(b"RIFF\x00\x00\x00\x00WAVE")  # bogus content; never read
    with pytest.raises(bz.UnsupportedError):
        await model.register_target_voice("speaker-01", ref_path)


@pytest.mark.skipif(
    not hasattr(bz, "VcModel")
    or os.getenv("BLAZEN_RUN_RVC_TESTS") != "1"
    or not os.getenv("BLAZEN_RVC_VOICE_ID")
    or not os.getenv("BLAZEN_RVC_INPUT_WAV"),
    reason="requires audio-vc-rvc feature, BLAZEN_RUN_RVC_TESTS=1, "
    "BLAZEN_RVC_VOICE_ID, and BLAZEN_RVC_INPUT_WAV "
    "(plus a voice profile under $BLAZEN_RVC_VOICE_DIR)",
)
async def test_rvc_convert_voice_live():
    """End-to-end live test against a pre-staged RVC voice profile."""
    voice_dir = os.getenv("BLAZEN_RVC_VOICE_DIR")
    voice_id = os.environ["BLAZEN_RVC_VOICE_ID"]
    input_wav = os.environ["BLAZEN_RVC_INPUT_WAV"]
    model = bz.VcModel.rvc(voice_dir=voice_dir) if voice_dir else bz.VcModel.rvc()
    wav_bytes = await model.convert_voice(input_wav, voice_id)
    assert isinstance(wav_bytes, (bytes, bytearray))
    assert len(wav_bytes) > 44  # at least a WAV header
    assert wav_bytes[:4] == b"RIFF"
    assert wav_bytes[8:12] == b"WAVE"
